### PR TITLE
claircore: bump to v0.2.0

### PR DESCRIPTION
**Pull request by ldelossa**
This commit patches ClairCore to v0.2.0.

The AffectedManifests API is updated to pass back a pointer
AffectedManifest struct.

The DeleteUpdateOperation API is updated to return the number of deleted
UpdateOperations.

All occurences of TestLog are replaced with zlog.Test

Various mocks scattered around tests have been consolidated to their
appropriate service packages and removed.

Signed-off-by: ldelossa <ldelossa@redhat.com>

### CODE REVIEW

File: indexer/mock.go [hunk 1]
```diff
@@ -9,25 +9,39 @@ import (
 var _ Service = (*Mock)(nil)
 
 // Mock implements a mock indexer Service
+//
+// Default Mock(s) are nil safe and return nil errors.
```
Comment (hdonnay) [inline indexer/mock.go]: Why do these all have nil checks now? I think I prefer the behavior where they'd panic the test if called accidentally.

File: notifier/service/testmode.go [hunk 2]
```diff
@@ -18,26 +18,26 @@ import (
...
 		if len(vulns) == 0 {
-			return claircore.NewAffectedManifests(), nil
+			return &claircore.AffectedManifests{}, nil
 		}
 
 		data := make([]byte, sha256.Size)
 		_, err := rand.Read(data)
 		if err != nil {
-			return claircore.AffectedManifests{}, err
+			return &claircore.AffectedManifests{}, err
```
Comment (hdonnay) [inline notifier/service/testmode.go]:
```suggestion
			return nil, err
```

File: notifier/service/testmode.go [hunk 3]
```diff
@@ -18,26 +18,26 @@ import (
...
 		}
 		digest, err := claircore.NewDigest("sha256", data)
 		if err != nil {
-			return claircore.AffectedManifests{}, err
+			return &claircore.AffectedManifests{}, err
```
Comment (hdonnay) [inline notifier/service/testmode.go]:
```suggestion
			return nil, err
```

File: httptransport/affectedmanifesthandler_test.go [hunk 4]
```diff
@@ -68,9 +57,9 @@ func testAffectedManifestHandlerIndexerOK(t *testing.T) {
 
 func testAffectedManifestHandlerIndexerErr(t *testing.T) {
 	t.Parallel()
-	h := AffectedManifestHandler(&affected{
-		affectedManifests: func(context.Context, []claircore.Vulnerability) (claircore.AffectedManifests, error) {
-			return claircore.NewAffectedManifests(), fmt.Errorf("error")
+	h := AffectedManifestHandler(&indexer.Mock{
+		AffectedManifests_: func(context.Context, []claircore.Vulnerability) (*claircore.AffectedManifests, error) {
+			return &claircore.AffectedManifests{}, fmt.Errorf("failed")
```
Comment (hdonnay) [inline httptransport/affectedmanifesthandler_test.go]:
```suggestion
			return nil, fmt.Errorf("failed")
```

File: httptransport/client/indexer.go [hunk 5]
```diff
@@ -26,29 +26,29 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
 		v,
 	})
 	if err != nil {
-		return affected, &clairerror.ErrBadVulnerabilities{err}
+		return &affected, &clairerror.ErrBadVulnerabilities{err}
```
Comment (hdonnay) [inline httptransport/client/indexer.go]:
```suggestion
		return nil, &clairerror.ErrBadVulnerabilities{err}
```

File: httptransport/client/indexer.go [hunk 6]
```diff
@@ -26,29 +26,29 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
...
 	}
 
 	u, err := s.addr.Parse(httptransport.AffectedManifestAPIPath)
 	if err != nil {
-		return affected, fmt.Errorf("failed to parse api address: %v", err)
+		return &affected, fmt.Errorf("failed to parse api address: %v", err)
```
Comment (hdonnay) [inline httptransport/client/indexer.go]:
```suggestion
		return nil, fmt.Errorf("failed to parse api address: %v", err)
```

File: httptransport/client/indexer.go [hunk 7]
```diff
@@ -26,29 +26,29 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
...
 	}
 	req, err := http.NewRequestWithContext(ctx, "POST", u.String(), buf)
 	if err != nil {
-		return affected, fmt.Errorf("failed to create request: %v", err)
+		return &affected, fmt.Errorf("failed to create request: %v", err)
```
Comment (hdonnay) [inline httptransport/client/indexer.go]:
```suggestion
		return nil, fmt.Errorf("failed to create request: %v", err)
```

File: httptransport/client/indexer.go [hunk 8]
```diff
@@ -26,29 +26,29 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
...
 	}
 	resp, err := s.c.Do(req)
 	if resp != nil {
 		defer resp.Body.Close()
 	}
 	if err != nil {
-		return affected, &clairerror.ErrRequestFail{Code: resp.StatusCode, Status: resp.Status}
+		return &affected, &clairerror.ErrRequestFail{Code: resp.StatusCode, Status: resp.Status}
```
Comment (hdonnay) [inline httptransport/client/indexer.go]:
```suggestion
		return nil, &clairerror.ErrRequestFail{Code: resp.StatusCode, Status: resp.Status}
```

File: httptransport/client/indexer.go [hunk 9]
```diff
@@ -26,29 +26,29 @@ func (s *HTTP) AffectedManifests(ctx context.Context, v []claircore.Vulnerabilit
...
 	}
 	err = json.NewDecoder(resp.Body).Decode(&affected)
 	if err != nil {
-		return affected, &clairerror.ErrBadAffectedManifests{err}
+		return &affected, &clairerror.ErrBadAffectedManifests{err}
```
Comment (hdonnay) [inline httptransport/client/indexer.go]:
```suggestion
		return nil, &clairerror.ErrBadAffectedManifests{err}
```

File: matcher/mock.go [hunk 10]
```diff
@@ -12,8 +12,10 @@ import (
 var _ Service = (*Mock)(nil)
 
 // Mock implements a mock matcher service
+//
+// All methods on a default Mock return nil safe return parameters and nil(s) for errors.
```
Comment (hdonnay) [inline matcher/mock.go]: Why do these all have nil checks? I think the panic behavior is better.

File: notifier/processor_create_test.go [hunk 11]
```diff
@@ -135,8 +135,8 @@ func testProcessorIndexerErr(t *testing.T) {
 		},
 	}
 	im := &indexer.Mock{
-		AffectedManifests_: func(ctx context.Context, vulns []claircore.Vulnerability) (claircore.AffectedManifests, error) {
-			return claircore.AffectedManifests{}, fmt.Errorf("expected")
+		AffectedManifests_: func(ctx context.Context, vulns []claircore.Vulnerability) (*claircore.AffectedManifests, error) {
+			return &claircore.AffectedManifests{}, fmt.Errorf("expected")
```
Comment (hdonnay) [inline notifier/processor_create_test.go]:
```suggestion
			return nil, fmt.Errorf("expected")
```

File: notifier/processor_create_test.go [hunk 12]
```diff
@@ -216,15 +216,15 @@ func testProcessorCreate(t *testing.T) {
...
 			case 0:
 				count++
-				return *affectedManifestsAdd, nil
+				return affectedManifestsAdd, nil
 			case 1:
-				return *affectedManifestsRemoved, nil
+				return affectedManifestsRemoved, nil
 			default:
-				return claircore.AffectedManifests{}, fmt.Errorf("unexpected number of calls")
+				return &claircore.AffectedManifests{}, fmt.Errorf("unexpected number of calls")
```
Comment (hdonnay) [inline notifier/processor_create_test.go]:
```suggestion
				return nil, fmt.Errorf("unexpected number of calls")
```

**Review by hdonnay — CHANGES_REQUESTED** [review]

File: indexer/mock.go (hunk 1 above)
Comment (ldelossa) [reply indexer/mock.go]: It allows for the default constructed mock to be useful in tests, instead of having to write the empty value returns often.

**Comment by ldelossa** [comment]
For your nil corrections, I don't think that is right.

It would be right to return a nil if an error is returned as well, but if we are returning an nil error, it would be painful to have any test usages have check the return valued for nil safety, that's not the idiom that's commonly used.

**Comment by ldelossa** [comment]
Ah, nvm, that's literally the changes you're suggesting. 

Think I was just rushing thru to fix compile errors initially. I'll fix.

**Review by hdonnay — APPROVED** [review]
lgtm

**Pull request merged**